File: src/library/grDevices/R/gradients.R
#  File src/library/grDevices/R/gradients.R
#  Part of the R package, https://www.R-project.org
#
#  Copyright (C) 2019      The R Foundation
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  A copy of the GNU General Public License is available at
#  https://www.R-project.org/Licenses/

## Create R objects defining gradients

extendOptions <- c("pad", "repeat", "reflect", "none")

linearGradient <- function(colours = c("black", "white"),
                           stops = seq(0, 1, length.out = length(colours)),
                           x1 = 0, y1 = 0, x2 = 1, y2 = 1,
                           extend = "pad") {
    ## Vectorising colours & stops
    nstops <- max(length(colours), length(stops))
    colours <- rep(colours, length.out = nstops)
    stops <- rep(stops, length.out = nstops)

    extend <- match(extend, extendOptions)
    if (is.na(extend))
        stop("Invalid 'extend' value")
    
    ## Make sure we really have colours in some form
    ## AND convert to "#RRGGBB" format
    colours <- rgb(t(col2rgb(colours)), maxColorValue=255)
    
    grad <- list(x1 = as.numeric(x1), y1 = as.numeric(y1),
                 x2 = as.numeric(x2), y2 = as.numeric(y2),
                 stops = as.numeric(stops), colours = colours,
                 extend = as.integer(extend))
    class(grad) <- c("LinearGradient", "Gradient")
    grad
}

File: src/main/gradients.c
/*
 *  R : A Computer Language for Statistical Data Analysis
 *  Copyright (C) 2019	     The R Foundation
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, a copy is available at
 *  https://www.R-project.org/Licenses/
 */

/* This should be regarded as part of the graphics engine */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <Defn.h>
#include <R_ext/GraphicsEngine.h>

/*
 * C API for graphics devices to interrogate gradient SEXPs
 */

#define gradient_x1 0
#define gradient_y1 1
#define gradient_x2 2
#define gradient_y2 3
#define gradient_stops 4
#define gradient_colours 5
#define gradient_extend 6

double R_GE_gradientX1(SEXP gradient)
{
    return REAL(VECTOR_ELT(gradient, gradient_x1))[0];
}

double R_GE_gradientY1(SEXP gradient)
{
    return REAL(VECTOR_ELT(gradient, gradient_y1))[0];
}

double R_GE_gradientX2(SEXP gradient)
{
    return REAL(VECTOR_ELT(gradient, gradient_x2))[0];
}

double R_GE_gradientY2(SEXP gradient)
{
    return REAL(VECTOR_ELT(gradient, gradient_y2))[0];
}

int R_GE_gradientNumStops(SEXP gradient) 
{
    return LENGTH(VECTOR_ELT(gradient, gradient_stops));
}

double R_GE_gradientStop(SEXP gradient, int i) 
{
    return REAL(VECTOR_ELT(gradient, gradient_stops))[i];
}

rcolor R_GE_gradientColour(SEXP gradient, int i) 
{
    return RGBpar(VECTOR_ELT(gradient, gradient_colours), i);
}

int R_GE_gradientExtend(SEXP gradient) 
{
    return INTEGER(VECTOR_ELT(gradient, gradient_extend))[0];
}
